p2p/src/secret_connection.rs

```diff
@@ -218,16 +220,49 @@ impl Handshake<AwaitingAuthSig> {
     }
 }
 
+struct SecretConnectionSendState {
+    send_nonce: Nonce,
+}
+
+impl Default for SecretConnectionSendState {
+    fn default() -> Self {
+        Self {
+            send_nonce: Nonce::default(),
+        }
+    }
+}
+
+struct SecretConnectionRecvState {
+    recv_nonce: Nonce,
+    recv_buffer: Vec<u8>,
+}
+
+impl Default for SecretConnectionRecvState {
+    fn default() -> Self {
+        Self {
+            recv_nonce: Nonce::default(),
+            recv_buffer: vec![],
+        }
+    }
+}
+
 /// Encrypted connection between peers in a Tendermint network.
-pub struct SecretConnection<IoHandler: Read + Write + Send + Sync> {
+///
+/// Can be safely shared between multiple threads, however only one thread can
+/// read from and one other thread can write to the secret connection
+/// simultaneously. This is intended to allow for use from two threads
+/// simultaneously: one thread for reading from the connection, and one thread
+/// for simultaneously writing to it.
+pub struct SecretConnection<IoHandler> {
     io_handler: IoHandler,
     protocol_version: Version,
-    recv_nonce: Nonce,
-    send_nonce: Nonce,
     recv_cipher: ChaCha20Poly1305,
     send_cipher: ChaCha20Poly1305,
     remote_pubkey: Option<PublicKey>,
-    recv_buffer: Vec<u8>,
+    // The portion of the connection state that is mutated when sending.
+    send_state: Arc<Mutex<SecretConnectionSendState>>,
+    // The portion of the connection state that is mutated when receiving.
+    recv_state: Arc<Mutex<SecretConnectionRecvState>>,
 }
```

Comment:
Why not just `Arc<Mutex<SecretConnection<_>>`?

Reply:
We could, but then it wouldn't be full-duplex, would it? (i.e. reading would block writing, and vice-versa).

I have two main assumptions here:

1. That `std::net::TcpStream` is full-duplex (I'm not 100% sure if it is though).
2. That the MConnection, which will make use of the `SecretConnection`, will need full-duplex communication (e.g. while we're receiving large chunks of data from a peer, we may need to write out to that peer simultaneously).

Reply:
Aah, okay, well in that case why not have a type for a half-duplex socket which only has a key for sending in one particular direction?

Reply:
So do you mean to have a `SecretConnectionSender` and a `SecretConnectionReceiver`?

Reply:
Part of the benefit I see of the approach I've implemented here is that it can be used in a single-threaded way, as is used in the KMS, which shouldn't change the existing communication pattern. But if the underlying `IoHandler` implements `TryClone` then you can have full-duplex communication in a multi-threaded way.

Reply:
You could still compose the full duplex version from the half-duplex versions. It'd require always cloning, but I don't think that should be a problem.

Reply:
Hmmm, yes, if we go that route we'd probably need to redo the tests for the `SecretConnection`, because they currently depend on this [pipe](https://github.com/arcnmx/pipe-rs) lib whose endpoints can't be cloned.

Reply:
Perhaps you could use a Unix domain socket instead?

Reply:
Perhaps you could use a Unix domain socket instead? (although I suppose that would preclude testing on Windows, at least those tests anyway)

Reply:
Not sure why TCP-based tests weren't used in the first place, but they seem to work just fine for our purposes (and act as decent integration tests too), so we may as well just use TCP-based tests right?

Reply:
Yeah TCP works just fine and is portable, the only tricky part is locating an unused port and handling the error case if something is bound to the test port

Reply:
Afaik Rust handles that automatically when you `bind` to `127.0.0.1:0`, doesn't it?